src/libraries/Common/src/SourceGenerators/GeneratorHelpers.cs

```diff
@@ -0,0 +1,12 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace SourceGenerators
+{
+    public static class GeneratorHelpers
+    {
+        public static string MakeNameUnique(ref string name) => name += $"_{new Random().Next():X8}";
```

Comment:
Not sure where this is being used, but `Random` doesn't guarantee uniqueness. If a naming conflict might compromise functional correctness or cause the generated code to fail to compile, may I suggest using a counter that gets incremented each time a suffix gets calculated using the `Interlocked.Increment` method.

Also, nit: the method is already returning the string so why take the argument by ref? It shouldn't have any perf benefits and it's better to keep simple:
```suggestion
        public static string MakeNameUnique(string name) => name += $"_{new Random().Next():X8}";
```

Reply:
I agree with Eirik, we used random in the options source gen because we wanted to generate suffix across multiple modules compilations which we can't use counter like what Eirik suggested.

Reply:
> If a naming conflict might compromise functional correctness or cause the generated code to fail to compile

This is not the case.

> We used random in the options source gen because we wanted to generate suffix across multiple modules compilations which we can't use counter like what Eirik suggested

I'm using `Random` for the same reason. Moved the logic into a shared abstraction with config binding so we do it the same way.

Reply:
> > If a naming conflict might compromise functional correctness or cause the generated code to fail to compile
> 
> This is not the case.

I saw that's it's being used to generate namespace names, so isn't it the case that if the same suffix is generated twice it could non-deterministically result in compiler errors?

> I'm using Random for the same reason. 

An important property of source generators and compilers in general is reproducible builds. In other words, equal inputs should deterministically produce equal outputs. If `Random` must be used, then I would strongly recommend passing a seed value.

Reply:
In any case, I believe compiler interception would point to the right interceptor, even if it's generated with the same name across assemblies. Seems like `file`-scoped emitted code would facilitate this characteristic. @RikkiGibson would you have any thoughts on this?

@captainsafia did you face these considerations in RDG?

Reply:
>  Seems like a Guid suffix might be a better approach. @tarekgh Tarek Mahmoud Sayed FTE was this considered as well?

Yes, I was trying to generate shorted suffix. For options, it is very unlikely we run into collision (the probability would be `2 / int.MaxValue`). Options will use suffix only if running with language version earlier than `11`.  The worst case if collision happen, recompile will fix it. Anyway, if want to use GUID, I am fine with that too. We'll end up with long names in the compiled code.

Reply:
I wouldn't consider `Random` to provide a perfectly uniform distribution over integers, but even if it was the birthday paradox suggests that the proability can increase substantially as you increase the number of samples. Unlike GUID generating conflicts with `Random` is trivial, consider this example:
```C#
var generated = new HashSet<int>();
for (int i = 0; ; i++)
{
    int random = new System.Random().Next();
    if (!generated.Add(random))
    {
        Console.WriteLine($"Found conflict after {i} iterations");
        break;
    }
}
```
I'm consistently getting a conflict below 50k iterations, but in some cases that can drop to as low as 1000.

>  The worst case if collision happen, recompile will fix it.

I don't think that would be acceptable behavior for product code that we ship. In my view the names should be generated deterministically in a way that avoid conflicts -- most compilers use incrementing counters to achieve this.

Reply:
For context here's the [article on randomness ](https://reproducible-builds.org/docs/randomness/) from the reproducible builds project page. 

cc @dotnet/roslyn in case they could point us to Roslyn-specific guidance for generating unique identifiers.

Reply:
> most compilers use incrementing counters to achieve this.

Using counters here is not possible as I indicated before we need to generate this across multiple compilations. 

> I'm consistently getting a conflict below 50k iterations, but in some cases that can drop to as low as 1000.

Still, this is very unlikely to happen considering other factors in the Options scenarios. The conflict can only have an effect if one assembly is internally visible to the other assembly, and both use options source generator. This is by itself is rare scenario. In addition, the users are using the language version earlier than `11`.  While we are requiring compiler ver `11` in other source generators I expect more users will opt-in using `11`. Doing the math on that, it is almost 0% for me. 

Anyway, if someone is still not comfortable with that, I am fine to go with using GUID.

Reply:
Just saw this conversation.  We definitely want to avoid random number generation in our source generators. As Eirik stated, it breaks determinism: build multiple times, get different results. And both we and customers depend on that determinism.  Plus @jaredpar will be at your virtual door the moment he finds out.

Reply:
We might want to have an analyzer that flags this API when used in a source generator.

Reply:
Why not just qualify the namespace name with the project name, or some hash of the project name? That would be deterministic and preferable to generating a random guid that changes on every build.

My concern is that we're adding a method in the common folder called `MakeNameUnique` which is random but doesn't guarantee uniqueness. Even though the use case at hand has a low (but still nonzero) probability of conflict, it could end up being adopted elsewhere where that isn't the case.

Reply:
Feel free to submit a PR for this Eirik.

Reply:
I'm going OOF so won't be able to track this. I would recommend either fixing it here or file an issue tracking removal of randomness from the source generators.

Reply:
> We might want to have an analyzer that flags this API when used in a source generator.

We have an AnalyzerBannedApis analyzer which uses the common banned API txt format. Can add it to that.

Reply:
> I'm going OOF so won't be able to track this. I would recommend either fixing it here or file an issue tracking removal of randomness from the source generators.

I'll handle the follow up here.